File: packages/upay-js/src/resources/PaymentLinks.ts
/**
 * Recurso de Payment Links
 */

import { HttpClient } from '../utils/http';
import type {
  CreatePaymentLinkRequest,
  UpdatePaymentLinkRequest,
  PaymentLink,
  PaginationParams,
  PaginatedResponse,
} from '../types';

export class PaymentLinksResource {
  constructor(private http: HttpClient) {}

  /**
   * Mapeia a resposta da API para um PaymentLink
   * @private
   */
  private mapPaymentLinkResponse(response: any): PaymentLink {
    return response.paymentLink || response.data || response;
  }

  /**
   * Cria um novo link de pagamento
   */
  async create(data: CreatePaymentLinkRequest): Promise<PaymentLink> {
    // Validação básica
    if (!data.title || data.title.trim().length < 3) {
      throw new Error('Título deve ter pelo menos 3 caracteres');
    }

    if (!data.amount && (!data.products || data.products.length === 0)) {
      throw new Error('É necessário fornecer amount ou products');
    }

    if (data.amount && data.amount < 100) {
      throw new Error('Valor mínimo é R$ 1,00 (100 centavos)');
    }

    const response = await this.http.post<any>('/payment-links', {
      title: data.title,
      description: data.description,
      amountCents: data.amount,
      products: data.products,
      currency: data.currency || 'BRL',
      expiresAt: data.expiresAt,
      redirectUrl: data.redirectUrl,
      settings: data.settings,
      status: data.status || 'ACTIVE',
      metaPixelCode: data.metaPixelCode,
      stockQuantity: data.stockQuantity,
      stockEnabled: data.stockEnabled,
    });
    
    return this.mapPaymentLinkResponse(response);
  }

  /**
   * Lista links de pagamento
   */
  async list(params?: PaginationParams & { status?: string }): Promise<PaginatedResponse<PaymentLink>> {
    const response = await this.http.get<any>('/payment-links', {
      page: params?.page,
      limit: params?.limit,
      cursor: params?.cursor,
      orderBy: params?.orderBy,
      orderDirection: params?.orderDirection,
      status: params?.status,
    });
    
    // Mapear resposta da API: { message, paymentLinks, pagination } -> { data, pagination }
    return {
      data: response.paymentLinks || response.data || [],
      pagination: response.pagination || { total: 0, page: 1, limit: 10 }
    };
  }

  /**
   * Obtém um link de pagamento por ID
   */
  async get(id: string): Promise<PaymentLink> {
    if (!id) {
      throw new Error('ID é obrigatório');
    }
    const response = await this.http.get<any>(`/payment-links/${id}`);
    return this.mapPaymentLinkResponse(response);
  }

  /**
   * Obtém um link de pagamento por slug
   */
  async getBySlug(slug: string): Promise<PaymentLink> {
    if (!slug) {
      throw new Error('Slug é obrigatório');
    }
    const response = await this.http.get<any>(`/payment-links/slug/${slug}`);
    return this.mapPaymentLinkResponse(response);
  }

  /**
   * Atualiza um link de pagamento
   */
  async update(id: string, data: UpdatePaymentLinkRequest): Promise<PaymentLink> {
    if (!id) {
      throw new Error('ID é obrigatório');
    }

    const updateData: any = {};
    if (data.title !== undefined) updateData.title = data.title;
    if (data.description !== undefined) updateData.description = data.description;
    if (data.amount !== undefined) updateData.amountCents = data.amount;
    if (data.status !== undefined) updateData.status = data.status;
    if (data.expiresAt !== undefined) updateData.expiresAt = data.expiresAt;
    if (data.redirectUrl !== undefined) updateData.redirectUrl = data.redirectUrl;
    if (data.settings !== undefined) updateData.settings = data.settings;

    const response = await this.http.patch<any>(`/payment-links/${id}`, updateData);
    return this.mapPaymentLinkResponse(response);
  }

  /**
   * Deleta um link de pagamento
   */
  async delete(id: string): Promise<void> {
    if (!id) {
      throw new Error('ID é obrigatório');
    }
    await this.http.delete(`/payment-links/${id}`);
  }

  /**
   * Obtém a URL pública do checkout
   */
  getCheckoutUrl(slug: string, baseUrl?: string): string {
    // Validar slug
    if (!slug || typeof slug !== 'string' || slug.trim().length === 0) {
      throw new Error('Slug é obrigatório e deve ser uma string não vazia');
    }

    // Normalizar baseUrl (remover trailing slash se existir)
    const normalizedBaseUrl = (baseUrl || 'https://checkout.upaybr.com').trim().replace(/\/+$/, '');
    
    // Normalizar slug (remover leading slashes, trim, e URL-encode)
    const normalizedSlug = slug.trim().replace(/^\/+/, '');
    const encodedSlug = encodeURIComponent(normalizedSlug);
    
    return `${normalizedBaseUrl}/${encodedSlug}`;
  }
}

File: packages/upay-python/upay/resources/payment_links.py
"""
Recurso de Payment Links
"""

from typing import Optional, Dict, Any
from urllib.parse import quote
from ..http import HttpClient


class PaymentLinksResource:
    """Recurso para gerenciar Payment Links"""
    
    def __init__(self, http: HttpClient):
        self.http = http
    
    def _normalize_payment_link_response(self, response: Dict[str, Any]) -> Dict[str, Any]:
        """
        Normaliza a resposta da API para retornar o PaymentLink de forma consistente
        
        Args:
            response: Resposta bruta da API
            
        Returns:
            PaymentLink normalizado
        """
        return response.get("paymentLink") or response.get("data") or response
    
    def create(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Cria um novo link de pagamento
        
        Args:
            data: Dados do link de pagamento
                - title: Título do link (obrigatório, min 3 caracteres)
                - amount: Valor em centavos (obrigatório se não houver products)
                - description: Descrição do link
                - currency: Moeda (padrão: BRL)
                - expiresAt: Data de expiração (ISO string)
                - redirectUrl: URL de redirecionamento
                - settings: Configurações de pagamento
                - status: Status (ACTIVE ou INACTIVE)
                - products: Lista de produtos
                
        Returns:
            Link de pagamento criado
        """
        # Validação básica
        title = data.get("title")
        if not isinstance(title, str) or len(title.strip()) < 3:
            raise ValueError("Título deve ter pelo menos 3 caracteres")
        
        if not data.get("amount") and not data.get("products"):
            raise ValueError("É necessário fornecer amount ou products")
        
        # Validação robusta de amount
        if "amount" in data:
            amount = data["amount"]
            # Verificar se é numérico
            if not isinstance(amount, (int, float)):
                raise ValueError("Amount deve ser um número")
            # Converter para int se for float
            amount = int(amount)
            # Verificar se não é negativo
            if amount < 0:
                raise ValueError("Valor não pode ser negativo")
            # Verificar valor mínimo
            if amount < 100:
                raise ValueError("Valor mínimo é R$ 1,00 (100 centavos)")
        
        # Prepara dados para envio
        request_data = {
            "title": data["title"],
            "description": data.get("description"),
            "amountCents": data.get("amount"),
            "products": data.get("products"),
            "currency": data.get("currency", "BRL"),
            "expiresAt": data.get("expiresAt"),
            "redirectUrl": data.get("redirectUrl"),
            "settings": data.get("settings"),
            "status": data.get("status", "ACTIVE"),
            "metaPixelCode": data.get("metaPixelCode"),
            "stockQuantity": data.get("stockQuantity"),
            "stockEnabled": data.get("stockEnabled"),
        }
        
        # Remove valores None
        request_data = {k: v for k, v in request_data.items() if v is not None}
        
        response = self.http.post("/payment-links", request_data)
        
        # Mapear resposta: { message, data } -> retornar data
        return self._normalize_payment_link_response(response)
    
    def list(
        self,
        page: Optional[int] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
        order_by: Optional[str] = None,
        order_direction: Optional[str] = None,
        status: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Lista links de pagamento
        
        Args:
            page: Número da página
            limit: Limite de itens por página
            cursor: Cursor para paginação
            order_by: Campo para ordenação
            order_direction: Direção da ordenação (ASC ou DESC)
            status: Filtrar por status
            
        Returns:
            Dicionário com 'data' (lista) e 'pagination'
        """
        params = {
            "page": page,
            "limit": limit,
            "cursor": cursor,
            "orderBy": order_by,
            "orderDirection": order_direction,
            "status": status,
        }
        
        response = self.http.get("/payment-links", params)
        
        # Mapear resposta: { message, paymentLinks, pagination } -> { data, pagination }
        return {
            "data": response.get("paymentLinks") or response.get("data") or [],
            "pagination": response.get("pagination") or {"total": 0, "page": 1, "limit": 10}
        }
    
    def get(self, link_id: str) -> Dict[str, Any]:
        """
        Obtém um link de pagamento por ID
        
        Args:
            link_id: ID do link de pagamento
            
        Returns:
            Link de pagamento
        """
        if not link_id:
            raise ValueError("ID é obrigatório")
        
        response = self.http.get(f"/payment-links/{link_id}")
        
        # Mapear resposta: { message, paymentLink } -> retornar paymentLink
        return self._normalize_payment_link_response(response)
    
    def get_by_slug(self, slug: str) -> Dict[str, Any]:
        """
        Obtém um link de pagamento por slug
        
        Args:
            slug: Slug do link de pagamento
            
        Returns:
            Link de pagamento
        """
        if not slug:
            raise ValueError("Slug é obrigatório")
        
        response = self.http.get(f"/payment-links/slug/{slug}")
        return self._normalize_payment_link_response(response)
    
    def update(self, link_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Atualiza um link de pagamento
        
        Args:
            link_id: ID do link de pagamento
            data: Dados para atualizar
            
        Returns:
            Link de pagamento atualizado
        """
        if not link_id:
            raise ValueError("ID é obrigatório")
        
        # Prepara dados para envio
        update_data = {}
        if "title" in data:
            update_data["title"] = data["title"]
        if "description" in data:
            update_data["description"] = data["description"]
        if "amount" in data:
            update_data["amountCents"] = data["amount"]
        if "status" in data:
            update_data["status"] = data["status"]
        if "expiresAt" in data:
            update_data["expiresAt"] = data["expiresAt"]
        if "redirectUrl" in data:
            update_data["redirectUrl"] = data["redirectUrl"]
        if "settings" in data:
            update_data["settings"] = data["settings"]
        
        response = self.http.patch(f"/payment-links/{link_id}", update_data)
        return self._normalize_payment_link_response(response)
    
    def delete(self, link_id: str) -> None:
        """
        Deleta um link de pagamento
        
        Args:
            link_id: ID do link de pagamento
        """
        if not link_id:
            raise ValueError("ID é obrigatório")
        
        self.http.delete(f"/payment-links/{link_id}")
    
    def get_checkout_url(self, slug: str, base_url: Optional[str] = None) -> str:
        """
        Obtém a URL pública do checkout
        
        Args:
            slug: Slug do link de pagamento
            base_url: URL base do checkout (padrão: https://checkout.upaybr.com)
            
        Returns:
            URL completa do checkout
        """
        # Validar slug
        if not slug or not isinstance(slug, str):
            raise ValueError("Slug é obrigatório e deve ser uma string")
        
        slug = slug.strip()
        if not slug:
            raise ValueError("Slug não pode ser vazio")
        
        # Normalizar base_url (remover trailing slash)
        checkout_base = (base_url or "https://checkout.upaybr.com").rstrip('/')
        
        # Codificar slug para URL
        encoded_slug = quote(slug, safe='')
        
        return f"{checkout_base}/{encoded_slug}"
